mountpoint-s3-crt/src/s3/buffer.rs

```diff
@@ -0,0 +1,136 @@
+#[derive(Debug)]
+pub struct Buffer<'slice> {
+    slice: &'slice aws_byte_cursor,
+    ticket: Option<&'slice BufferTicket>,
+}
+
+impl<'slice> Buffer<'slice> {
+    /// Create a new instance from a [aws_byte_cursor] and optional associated [aws_s3_buffer_ticket].
+    ///
+    /// SAFETY: slice must be a valid `aws_byte_cursor`.
+    pub(crate) fn new_unchecked(slice: &'slice aws_byte_cursor, ticket: Option<&'slice BufferTicket>) -> Self {
+        Self { slice, ticket }
+    }
+
+    /// Acquire ownership of the underlying memory pool buffer and return an [OwnedBuffer].
+    ///
+    /// Fails and returns [None] if no [BufferTicket] was provided.
+    pub fn to_owned_buffer(&self) -> Option<OwnedBuffer> {
+        let ticket = self.ticket?.clone();
+        Some(OwnedBuffer::new(*self.slice, ticket))
+    }
+}
+
+impl Deref for Buffer<'_> {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        // SAFETY: `self.slice` is valid for the lifetime of self.
+        unsafe { aws_byte_cursor_as_slice(self.slice) }
+    }
+}
+
+impl AsRef<[u8]> for Buffer<'_> {
+    fn as_ref(&self) -> &[u8] {
+        self.deref()
+    }
+}
+
+/// An owned data buffer from the memory pool.
+#[derive(Debug, Clone)]
+pub struct OwnedBuffer {
+    slice: aws_byte_cursor,
+    _ticket: BufferTicket,
+}
+
+impl OwnedBuffer {
+    /// Create a new instance.
+    fn new(slice: aws_byte_cursor, ticket: BufferTicket) -> Self {
+        Self { slice, _ticket: ticket }
+    }
+}
+
+// SAFETY: `self.slice` is valid for the lifetime of `self._ticket`
+unsafe impl Send for OwnedBuffer {}
+// SAFETY: `self.slice` is valid for the lifetime of `self._ticket`
+unsafe impl Sync for OwnedBuffer {}
+
+impl Deref for OwnedBuffer {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        // SAFETY: `self.slice` is valid as for the lifetime of `self.ticket`.
+        unsafe { aws_byte_cursor_as_slice(&self.slice) }
+    }
+}
+
+impl AsRef<[u8]> for OwnedBuffer {
+    fn as_ref(&self) -> &[u8] {
+        self.deref()
+    }
+}
+
+/// Wraps a valid [aws_s3_buffer_ticket].
+#[derive(Debug)]
+pub(crate) struct BufferTicket {
+    inner: NonNull<aws_s3_buffer_ticket>,
+}
+
+impl BufferTicket {
+    /// Create a new instance.
+    ///
+    /// # Safety
+    /// `ticket` must be a valid `aws_s3_buffer_ticket`.
+    pub unsafe fn new(ticket: NonNull<aws_s3_buffer_ticket>) -> Self {
+        aws_s3_buffer_ticket_acquire(ticket.as_ptr());
+        Self { inner: ticket }
+    }
+
+    /// Create an instance without incrementing the reference count of the underlying ticket.
+    /// Returns `None` if `ticket` is `null`.
+    ///
+    /// # Safety
+    /// `ticket` must be valid for the lifetime of the returned `Self`.
+    pub unsafe fn new_unowned(ticket: *mut aws_s3_buffer_ticket) -> Option<ManuallyDrop<Self>> {
+        let inner = NonNull::new(ticket)?;
+        Some(ManuallyDrop::new(Self { inner }))
```

Comment:
is `ManuallyDrop` a good fit here? caller of `new_unowned` may manually drop the returned value, while we haven't called `aws_s3_buffer_ticket_acquire`

may be `drop` could handle itself? e.g. based on a boolean flag:
```
pub(crate) struct BufferTicket {
    inner: NonNull<aws_s3_buffer_ticket>,
    owned: bool,
}
```

Reply:
You are right, the "unowned" part is confusing. I've reworked `Buffer` to hold the `aws_s3_buffer_ticket` pointer directly and only build a `BufferTicket` wrapper when acquiring ownership.